File: computational_physics/Homework8/hw8.jl
using Random
using Plots
using Statistics
function f(x)
    return x^2
end

function W(x)
    return x^2 + 2.0
end

function W_normalized(x, N)
    return W(x) / N
end

function estimate_normalization_constant(M, a, b)
    total_weight = 0.0
    for _ in 1:M
        x = a + (b - a) * rand() 
        weight = W(x)
        total_weight += weight
    end
    N = (b - a) * total_weight / M
    return N
end

function monte_carlo_integration(a, b, M, N)
    integral_sum = 0.0
    sample_sum = 0.0
    total_weight = 0.0
    autocorrelation_values = Float64[]
    sample_squared_sum = 0.0

    for i in 1:(M + 80000)
        x = rand()  # Generate a random number in [0, 1]
        # print(x)
        weight_0 = W_normalized(x, N)
        rand_val = (2*rand()-1.0) *1.27
        x_next = (b - a) * rand_val + x
        weight_1 = W_normalized(x_next, N)
        prob = weight_1 / weight_0
        si = rand()
        
        if prob >= si && x_next <= 1.0
            x = x_next
        end

        integrand_val = f(x)
        push!(autocorrelation_values, integrand_val / W_normalized(x, N))

        if i >= 80000
            integral_sum += integrand_val / W_normalized(x, N)
            sample_sum += integrand_val / W_normalized(x, N)
            sample_squared_sum += integrand_val^2
            total_weight += W_normalized(x, N)
        end
    end

    integral_estimate = integral_sum / M
    sample_mean = sample_sum / M
    sample_mean_squared = sample_squared_sum / M
    sample_mean_stddev = sqrt((sample_mean_squared - sample_mean^2) / M)

    return integral_estimate, sample_mean, sample_mean_stddev, autocorrelation_values
end

function autocorrelation(sample_points, lag)
    M = length(sample_points)
    if lag >= M
        return 0.0
    end

    mean_sample = mean(sample_points)
    numerator = sum([sample_points[i] * sample_points[i + lag] for i in 1:(M - lag)]) / M
    denominator = var(sample_points)

    if denominator == 0.0
        return 0.0
    else
        return (numerator - mean_sample^2) / denominator
    end
end

a = 0.0  # Lower limit of the integration interval
b = 1.0  # Upper limit of the integration interval
M = 100000  # Number of sample points
N = estimate_normalization_constant(M, a, b)
println("Normalization Constant: $N")
estimated_integral, mean_sample, sample_mean_stddev, autocorrelation_values = monte_carlo_integration(a, b, M, N)
println("Monte Carlo Integral Estimate with Importance Sampling: $estimated_integral")
println("Sample Mean: $mean_sample")
println("Standard Deviation of Sample Mean: $sample_mean_stddev")

max_lag = 89000  
autocorrelation_l = [autocorrelation(autocorrelation_values, lag) for lag in 1:max_lag]

# Plot the autocorrelation function
plot(autocorrelation_l, xlabel="Lag", ylabel="Autocorrelation", title="Autocorrelation Function", legend=false, grid=true)

File: computational_physics/Homework8/hw_8.py
import random
import numpy as np
import matplotlib.pyplot as plt

def W(x):
    return (x**2+2.0)

M = 10000# Number of sample points
N = 0.0  # Initialize normalization constant

# Determine the normalization constant N by integrating W(x) over the range [0, 1]
for _ in range(M):
    x = random.uniform(0, 1)
    N += W(x)
N /= M

def integrand(x):
    return x**2

# Perform the Monte Carlo integration with importance sampling
def monte_carlo_integral(M, N):
    integral_sum = 0.0
    total_weight = 0.0  
    sample_means = []
    sample_points = []
    total_points = []

    for i in range(M + 80000):
        x = random.uniform(0, 1)
        weight = W(x) /N  # Importance sampling weight
        integrand_val = integrand(x)
        total_points.append(integrand_val/weight)
        if i >= 80000:
            sample_points.append(integrand_val)
            integral_sum += integrand_val 
            total_weight += weight
            if i >= 80000 and (i + 1) % 100 == 0:
            # Calculate the sample mean for every 100 points
                sample_mean = integral_sum / total_weight
                sample_means.append(sample_mean)
    # Calculate the integral result
    integral_result = integral_sum / total_weight

    return integral_result, sample_means, sample_points,total_points

# Calculate the integral, sample means, and sample points
integral_result, sample_means, sample_points ,total_points= monte_carlo_integral(M, N)

# Calculate the standard deviation of the sample mean
standard_deviation = np.std(sample_means)
print("Monte Carlo Integral Result:", integral_result)
print("Standard Deviation of Sample Mean:", standard_deviation)


# \begin{array}{l}
# C(l)=\frac{\left\langle A_{n+l} A_n\right\rangle\left\langle A_n\right\rangle^2}{\left\langle A_n{ }^2\right\rangle-\left\langle A_n\right\rangle^2} \\
# \text { where } \left\langle A_{n+l} A_n\right\rangle=\frac{1}{M} \sum_{n=1}^M A_{n+1} A_n .
# \end{array}
import numpy as np
import matplotlib.pyplot as plt

# Autocorrelation function calculation
def autocorrelation(sample_points, lag):
    M = len(sample_points)
    if lag >= M:
        return 0.0

    mean_sample = np.mean(sample_points)
    numerator = sum((sample_points[i] * sample_points[i + lag] for i in range(M - lag))) / M
    denominator = np.var(sample_points)
    
    if denominator == 0:
        return 0.0
    else:
        return (numerator - mean_sample * mean_sample) / denominator

# Calculate autocorrelation values for various lags
max_lag = 89000  
autocorrelation_values = [autocorrelation(total_points, lag) for lag in range(max_lag)]

# Plot the autocorrelation function
plt.figure(figsize=(10, 5))
plt.plot(autocorrelation_values)
plt.title("Autocorrelation Function")
plt.xlabel("Lag")
plt.ylabel("Autocorrelation")
plt.grid(True)
plt.show()



